test/radix/quic_bindings.c

```diff
@@ -0,0 +1,788 @@
+ * not assume a single client and a single server. Examples of vignettes
+ * designed to be supported by the radix test framework in future include:
+ *
+ *      single client    <-> single server
+ *      multiple clients <-> single server
+ *      single client    <-> multiple servers
+ *      multiple clients <-> multiple servers
+ *
+ * 'Multi-process' and 'multi-node' means there has been some consideration
+ * given to support of multi-process and multi-node testing in the future,
+ * though this is not currently supported.
+ */
+
+/*
+ * An object is something associated with a name in the process-level state. The
+ * process-level state primarily revolves around a global dictionary of SSL
+ * objects.
+ */
+typedef struct radix_obj_st {
+    char                *name;  /* owned, zero-terminated */
+    SSL                 *ssl;   /* owns one reference */
+    unsigned int        registered      : 1; /* in LHASH? */
+    unsigned int        active          : 1; /* tick? */
+} RADIX_OBJ;
+
+DEFINE_LHASH_OF_EX(RADIX_OBJ);
+
+/* Process-level state (i.e. "globals" in the normal sense of the word) */
+typedef struct radix_process_st {
+    size_t                  node_idx;
+    size_t                  process_idx;
+    size_t                  next_thread_idx;
+    STACK_OF(RADIX_THREAD)  *threads;
+
+    /* Process-global state. */
+    CRYPTO_MUTEX            *gm;            /* global mutex */
+    LHASH_OF(RADIX_OBJ)     *objs;          /* protected by gm */
+    OSSL_TIME               time_slip;      /* protected by gm */
+    BIO                     *keylog_out;    /* protected by gm */
+
+    int                     done_join_all_threads;
+
+    /*
+     * Valid if done_join_all threads. Logical AND of all child worker results.
+     */
+    int                     thread_composite_testresult;
+} RADIX_PROCESS;
+
+#define NUM_SLOTS       4
+
+/* Thread-level state within a process */
+typedef struct radix_thread_st {
+    RADIX_PROCESS       *rp;
+    CRYPTO_THREAD       *t;
+    unsigned char       *tmp_buf;
+    size_t              tmp_buf_offset;
+    size_t              thread_idx; /* 0=main thread */
+    RADIX_OBJ           *slot[NUM_SLOTS];
+    SSL                 *ssl[NUM_SLOTS];
+
+    /* child thread spawn arguments */
+    SCRIPT_INFO         *child_script_info;
+    BIO                 *debug_bio;
+
+    /* m protects all of the below values */
+    CRYPTO_MUTEX        *m;
+    int                 done;
+    int                 testresult; /* valid if done */
+
+    uint64_t            scratch0;
+} RADIX_THREAD;
+
+DEFINE_STACK_OF(RADIX_THREAD)
+
+/* ssl reference is transferred. name is copied and is required. */
+static RADIX_OBJ *RADIX_OBJ_new(const char *name, SSL *ssl)
+{
+    RADIX_OBJ *obj;
+
+    if (!TEST_ptr(name) || !TEST_ptr(ssl))
+        return NULL;
+
+    if (!TEST_ptr(obj = OPENSSL_zalloc(sizeof(*obj))))
+       return NULL;
+
+    obj->name = OPENSSL_strdup(name);
```

Comment:
no `if (!TEST_PTR(obj->name = OPENSSL_strdup(name))` check here? is that intentional?